### PR TITLE
Carry over hash value for all literal types in remote caching

### PR DESCRIPTION
**Pull request by nicholasjng**
# TL;DR
Previously, the hash on an input literal would be ignored if the literal was of either the `Collection` or `Map` type. This is fixed here by moving the hash-checking branch up in front of the special type logic, so that an empty literal with hash is constructed in either case if a hash is found.

Adds a regression test to verify the hashes are carried over into the representer literals for both map and collection types.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
See the discussion in https://github.com/flyteorg/flytekit/pull/1720 for the problem statement. This PR is a direct follow-up.

The remote cache fix was missing from that PR, and that is exactly what is added here. The code here is the same fix approach, just in Go this time.

## Tracking Issue
N/A

### CODE REVIEW

File: go/tasks/pluginmachinery/catalog/hashing_test.go [hunk 1]
```diff
@@ -593,3 +593,33 @@ func TestNoInputValues(t *testing.T) {
 	assert.Equal(t, "GKw-c0PwFokMUQ6T-TUmEWnZ4_VlQ2Qpgw-vCTT0-OQ", hashDupe)
 	assert.Equal(t, hashDupe, hash)
 }
+
+// Ensure that hash values on maps and collections are carried over
+// into representing literals.
+func TestHashCarryOnContainers(t *testing.T) {
```
Comment (eapolinario) [inline go/tasks/pluginmachinery/catalog/hashing_test.go:599]: can you move these tests to the table-tests in `TestHashLiteralMap_LiteralsWithHashSet` instead?

Comment (nicholasjng) [reply go/tasks/pluginmachinery/catalog/hashing_test.go]: Done in f4d830ee7b44b30399082eaaadb069e740fd2ebb.
